### PR TITLE
More bls

### PR DESCRIPTION
**Pull request by david-stolp**

### CODE REVIEW

File: core/primitives/src/aggregate_signature.rs [hunk 1]
```diff
@@ -84,10 +87,22 @@ impl<E: Engine> SecretKey<E> {
 }
 
 impl<E: Engine> PublicKey<E> {
+    pub fn empty() -> Self {
+        PublicKey{ point: E::G1Affine::zero() }
```
Comment (bowenwang1996) [inline core/primitives/src/aggregate_signature.rs]: space?

File: core/primitives/src/aggregate_signature.rs [hunk 2]
```diff
@@ -154,63 +178,86 @@ impl<E: Engine> From<&SecretKey<E>> for Vec<u8> {
...
+    fn from_bytes(v: Vec<u8>) -> Result<Self, Box<Error>> {
         let expected = <<E::G1Affine as CurveAffine>::Compressed as EncodedPoint>::size();
         if v.len() != expected {
-            panic!("invalid signature");
+            Err(LengthError(expected, v.len()))?;
         }
         let mut compressed = CompressedPublicKey::empty();
         compressed.as_mut().copy_from_slice(v.as_ref());
-        compressed.decompress().unwrap()
+        Ok(compressed.decompress()?)
```
Comment (bowenwang1996) [inline core/primitives/src/aggregate_signature.rs]: Isn't this equivalent to `compressed.decompress()`?

File: core/primitives/src/aggregate_signature.rs [hunk 3]
```diff
@@ -154,63 +178,86 @@ impl<E: Engine> From<&SecretKey<E>> for Vec<u8> {
...
+        write!(f, "{}, expected {}, got {}", self.description(), self.0, self.1)
+    }
+}
+
+impl<E: Engine> FromBytes for PublicKey<E> {
+    fn from_bytes(v: Vec<u8>) -> Result<Self, Box<Error>> {
         let expected = <<E::G1Affine as CurveAffine>::Compressed as EncodedPoint>::size();
         if v.len() != expected {
-            panic!("invalid signature");
+            Err(LengthError(expected, v.len()))?;
```
Comment (bowenwang1996) [inline core/primitives/src/aggregate_signature.rs]: This is obscure. It's better to write `return Err(LengthError(expected, v.len()));`

File: core/primitives/src/aggregate_signature.rs [hunk 4]
```diff
@@ -154,63 +178,86 @@ impl<E: Engine> From<&SecretKey<E>> for Vec<u8> {
...
 
-// TODO(#502): TryFrom instead of From since conversion can fail
-impl<E: Engine> From<Vec<u8>> for Signature<E> {
-    fn from(v: Vec<u8>) -> Self {
+impl<E: Engine> FromBytes for Signature<E> {
+    fn from_bytes(v: Vec<u8>) -> Result<Self, Box<std::error::Error>> {
         let expected = <<E::G2Affine as CurveAffine>::Compressed as EncodedPoint>::size();
         if v.len() != expected {
-            panic!("invalid signature");
+            Err(LengthError(expected, v.len()))?;
```
Comment (bowenwang1996) [inline core/primitives/src/aggregate_signature.rs]: Same here

File: core/primitives/src/types.rs [hunk 5]
```diff
@@ -70,6 +67,23 @@ impl GroupSignature {
...
+        if keys.len() < self.authority_mask.len() {
+            return false;
+        }
+        // Empty signature + empty public key would pass verification
+        if self.authority_count() == 0 {
+            return false;
+        }
+        let mut group_key = BlsAggregatePublicKey::new();
+        for (index, key) in keys.iter().enumerate() {
+            if self.authority_mask.get(index) == Some(&true) {
```
Comment (bowenwang1996) [inline core/primitives/src/types.rs]: `if let Some(true) = self.authority_mask.get(index)`

File: core/primitives/src/aggregate_signature.rs (hunk 2 above)
Comment (david-stolp) [reply core/primitives/src/aggregate_signature.rs]: No.  The Result's Err type is different.

File: core/primitives/src/aggregate_signature.rs (hunk 3 above)
Comment (david-stolp) [reply core/primitives/src/aggregate_signature.rs]: That does not compile.

Comment (bowenwang1996) [reply core/primitives/src/aggregate_signature.rs]: I see. I think explicit use of `into` might be better

**Pull request merged**